app/src/main/cpp/skyline/soc/gm20b/engines/maxwell_dma.cpp

```diff
@@ -30,167 +31,241 @@ namespace skyline::soc::gm20b::engine {
+                if (copyCache.size() < (srcSize + dstSize))
+                    copyCache.resize(srcSize + dstSize);
 
-        if (*registers.lineLengthIn != registers.dstSurface->width)
-            Logger::Warn("DMA copy width mismatch: src: {} dst: {}", *registers.lineLengthIn, registers.dstSurface->width);
+                dst = copyCache.data() + srcSize;
+            } else {
+                if (copyCache.size() < dstSize)
+                    copyCache.resize(dstSize);
 
-        gpu::texture::Dimensions srcDimensions{*registers.lineLengthIn, *registers.lineCount, 1};
-        size_t srcStride{srcDimensions.width * srcDimensions.height * bytesPerPixel};
+                dst = copyCache.data();
+            }
 
-        auto srcMappings{channelCtx.asCtx->gmmu.TranslateRange(*registers.offsetIn, srcStride)};
-        if (srcMappings.size() != 1) {
-            Logger::Warn("DMA for split textures is unimplemented");
-            return;
+            // If the destination is not entirely filled by the copy we copy it's current state in the cache to prevent clearing of other data.
+            if (registers.launchDma->dstMemoryLayout == Registers::LaunchDma::MemoryLayout::BlockLinear)
+                channelCtx.asCtx->gmmu.Read(dst, u64{*registers.offsetOut}, dstSize);
         }
 
-        gpu::texture::Dimensions dstDimensions{registers.dstSurface->width, registers.dstSurface->height, registers.dstSurface->depth};
-        dstDimensions.width = *registers.lineLengthIn; // We do not support copying subrects so we need the width to match on the source and destination
-        size_t dstBlockHeight{registers.dstSurface->blockSize.Height()}, dstBlockDepth{registers.dstSurface->blockSize.Depth()};
-        size_t dstLayerStride{gpu::texture::GetBlockLinearLayerSize(dstDimensions, 1, 1, bytesPerPixel, dstBlockHeight, dstBlockDepth)};
-
-        size_t dstLayerAddress{*registers.offsetOut + (registers.dstSurface->layer * dstLayerStride)};
-        auto dstMappings{channelCtx.asCtx->gmmu.TranslateRange(dstLayerAddress, dstLayerStride)};
-        if (dstMappings.size() != 1) {
-            Logger::Warn("DMA for split textures is unimplemented");
-            return;
-        }
+        copyCallback(src, dst);
 
-        Logger::Debug("{}x{}@0x{:X} -> {}x{}@0x{:X}", srcDimensions.width, srcDimensions.height, u64{*registers.offsetIn}, dstDimensions.width, dstDimensions.height, dstLayerAddress);
+        if (dstMappings.size() != 1)
+            channelCtx.asCtx->gmmu.Write(u64{*registers.offsetOut}, dst, dstSize);
 
-        gpu::texture::CopyLinearToBlockLinear(
-            dstDimensions,
-            1, 1, bytesPerPixel,
-            dstBlockHeight, dstBlockDepth,
-            srcMappings.front().data(), dstMappings.front().data()
-        );
+        // If the cache is over 5 MBs large then we clamp it to not waste memory
+        if (copyCache.size() > 5242880) [[unlikely]]
+            copyCache.resize(5242880);
```

Comment:
This won't do anything without calling shrink_to_fit, but is this a realistic case? It might be worth dropping this